packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -0,0 +1,287 @@
+            autoReconnect,
+            opsBehind,
+            online: OnlineStatus[isOnline()],
+            lastVisible: this.container.lastVisible !== undefined ?
+                performance.now() - this.container.lastVisible : undefined,
+            checkpointSequenceNumber,
+            sequenceNumber,
+        });
+
+        if (value === ConnectionState.Connected) {
+            this.firstConnection = false;
+            this.manualReconnectInProgress = false;
+        }
+    }
+
+    private setConnectionState(value: ConnectionState.Disconnected, reason: string);
+    private setConnectionState(value: ConnectionState.Connecting | ConnectionState.Connected);
+    private setConnectionState(
+        value: ConnectionState,
+        reason?: string,
+    ) {
+        assert(value !== ConnectionState.Connecting, "Trying to set connection state while container is connecting!");
+        if (this.connectionState === value) {
+            // Already in the desired state - exit early
+            this.logger.sendErrorEvent({ eventName: "setConnectionStateSame", value });
+            return;
+        }
+
+        const oldState = this._connectionState;
+        this._connectionState = value;
+
+        if (value === ConnectionState.Connected) {
+            // Mark our old client should have left in the quorum if it's still there
+            if (this._clientId !== undefined) {
+                const client: ILocalSequencedClient | undefined =
+                    this.protocolHandler?.quorum.getMember(this._clientId);
+                if (client !== undefined) {
+                    client.shouldHaveLeft = true;
+                }
+            }
+            this._clientId = this.pendingClientId;
+        } else if (value === ConnectionState.Disconnected) {
+            // Important as we process our own joinSession message through delta request
+            this.pendingClientId = undefined;
+        }
+
+        if (this.container.isLoaded) {
+            this.propagateConnectionState();
+        }
+
+        // Report telemetry after we set client id!
+        this.logConnectionStateChangeTelemetry(value, oldState, reason);
+    }
+
+    public propagateConnectionState() {
+        const logOpsOnReconnect: boolean =
+            this._connectionState === ConnectionState.Connected &&
+            !this.firstConnection &&
+            this.deltaManager.connectionMode === "write";
+        if (logOpsOnReconnect) {
+            this._messageCountAfterDisconnection = 0;
+        }
+
+        const state = this._connectionState === ConnectionState.Connected;
+        if (!this.context.disposed) {
+            this.context.setConnectionState(state, this.clientId);
+        }
+        assert(this.protocolHandler !== undefined, "Protocol handler should be set here");
+        this.protocolHandler.quorum.setConnectionState(state, this.clientId);
+        raiseConnectedEvent(this.logger, this.container, state, this.clientId);
```

Comment:
This is not a good pattern (of raising events on another class).
Can we take all of this logic (starting with line 275) and refactor into a callback that is passed in by container? Maybe whole that function?
Such that this class does not need to know about this.context, this.container?
I think it will need to know about this.protocolHandler as there is too much tied to it, but other things can be abstracted away, i think.
Ideally this class only about managing transitions, and nothing else, i.e. core of the logic that determines connectivity transition states (and probably tracking of clientIDs). Everything else ideally stays where it is.

Reply:
like above, could this class just raise an event here?

---
In reply to: [595601909](https://github.com/microsoft/FluidFramework/pull/5542#discussion_r595601909) [](ancestors = 595601909)